From 5a6a196fcab230413d4bfe8c56b3495df29cde79 Mon Sep 17 00:00:00 2001
From: Wilson Ding <dingwei@marvell.com>
Date: Tue, 3 Jul 2018 09:31:27 +0800
Subject: [PATCH 0663/1051] fix: pci: advk: Increase PIO timeout

The PIO transaction is mostly completed in 1 msec.
However, PIO transaction sometimes taks much more
time than we expected. In worse cases, it requires
24 msec for PIO done. In the meanwhile, Linux PCIe
core driver doesn't always check the return value
on pci_bus_read/write_config_xxx APIs. Eventually,
an incompleted PIO transaction following by another
one causes asynchronized abort on PCIe bus as below.
So let's set the PIO timeout value to 30 msec.

[    2.914196] advk-pcie d0070000.pcie: config read/write timed out
[    2.920360] Bad mode in Error handler detected on CPU0, code 0xbf000002 -- SError
[    2.928009] Internal error: Oops - bad mode: 0 [#1] PREEMPT SMP
[    2.934100] Modules linked in:
[    2.937241] CPU: 0 PID: 1 Comm: swapper/0 Not tainted 4.14.22-devel-18.05.0-00430-gc9b7e78 #4
[    2.946017] Hardware name: Marvell Armada 3720 Development Board DB-88F3720-DDR3 (DT)
[    2.954081] task: ffff80000e900000 task.stack: ffff000008068000
[    2.960183] PC is at advk_pcie_rd_conf+0x44/0x1d0
[    2.965017] LR is at pci_bus_read_config_word+0x68/0xa0
[    2.970388] pc : [<ffff0000085733fc>] lr : [<ffff000008549db8>] pstate: 400000c5
[    2.978003] sp : ffff00000806bbe0
[    2.981408] x29: ffff00000806bbe0 x28: ffff80000df3d2b8
[    2.986873] x27: 0000000000000005 x26: ffff00000906dc88
[    2.992339] x25: ffff000008bdf000 x24: 0000000000000040
[    2.997804] x23: 0000000000000000 x22: ffff80000e1f3398
[    3.003270] x21: ffff00000806bc6c x20: 0000000000000002
[    3.008736] x19: 0000000000000006 x18: 0000000000000010
[    3.014201] x17: 000000007f1f9caa x16: 000000005ff81112
[    3.019667] x15: 00000000ffffffff x14: ffff0000890464af
[    3.025133] x13: ffff0000090464bd x12: ffff000008f89df0
[    3.030598] x11: ffff000008648718 x10: ffff00000806b8a0
[    3.036064] x9 : 000000000000000d x8 : 2f77726974652074
[    3.041529] x7 : 6669672072656164 x6 : 0000000000000113
[    3.046995] x5 : ffff0000085733b8 x4 : 000000000000401c
[    3.052460] x3 : 0000000000000002 x2 : ffff00000a3c401c
[    3.057926] x1 : 0000000000000000 x0 : ffff80000e1f3800
[    3.063393] Process swapper/0 (pid: 1, stack limit = 0xffff000008068000)
[    3.070292] Call trace:
[    3.072802] Exception stack(0xffff00000806baa0 to 0xffff00000806bbe0)
[    3.079435] baa0: ffff80000e1f3800 0000000000000000 ffff00000a3c401c 0000000000000002
[    3.087498] bac0: 000000000000401c ffff0000085733b8 0000000000000113 6669672072656164
[    3.095562] bae0: 2f77726974652074 000000000000000d ffff00000806b8a0 ffff000008648718
[    3.103626] bb00: ffff000008f89df0 ffff0000090464bd ffff0000890464af 00000000ffffffff
[    3.111690] bb20: 000000005ff81112 000000007f1f9caa 0000000000000010 0000000000000006
[    3.119754] bb40: 0000000000000002 ffff00000806bc6c ffff80000e1f3398 0000000000000000
[    3.127818] bb60: 0000000000000040 ffff000008bdf000 ffff00000906dc88 0000000000000005
[    3.135882] bb80: ffff80000df3d2b8 ffff00000806bbe0 ffff000008549db8 ffff00000806bbe0
[    3.143946] bba0: ffff0000085733fc 00000000400000c5 ffff000008f89df0 ffff0000090464bd
[    3.152010] bbc0: ffffffffffffffff 00000000ffffffff ffff00000806bbe0 ffff0000085733fc
[    3.160076] [<ffff0000085733fc>] advk_pcie_rd_conf+0x44/0x1d0
[    3.165988] [<ffff000008549db8>] pci_bus_read_config_word+0x68/0xa0
[    3.172441] [<ffff000008551934>] __pci_bus_find_cap_start+0x1c/0x58
[    3.178890] [<ffff000008551994>] pci_find_capability+0x24/0x58
[    3.184893] [<ffff0000085531b4>] pci_af_flr+0x1c/0xd0
[    3.190092] [<ffff000008555af0>] pci_probe_reset_function+0xa0/0xd0
[    3.196542] [<ffff00000855a840>] pci_create_sysfs_dev_files+0x1f0/0x3a0
[    3.203355] [<ffff000008f10d18>] pci_sysfs_init+0x38/0x60
[    3.208908] [<ffff000008083250>] do_one_initcall+0x38/0x120
[    3.214643] [<ffff000008ef0d00>] kernel_init_freeable+0x188/0x22c
[    3.220913] [<ffff000008b5d0c0>] kernel_init+0x10/0x100
[    3.226288] [<ffff000008084480>] ret_from_fork+0x10/0x18
[    3.231756] Code: 840388d2 4200048b 5f0000b9 9f3e03d5 (c50640f9)

Change-Id: I116e3134fe1a0cd157391114f74807fb7df3490e
Signed-off-by: Wilson Ding <dingwei@marvell.com>
[Kevin: The original patch got from Marvell sdk10.0_19.06]
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/pci/controller/pci-aardvark.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/pci/controller/pci-aardvark.c b/drivers/pci/controller/pci-aardvark.c
index 6b4555ff2548..54f3b80e78b5 100644
--- a/drivers/pci/controller/pci-aardvark.c
+++ b/drivers/pci/controller/pci-aardvark.c
@@ -166,7 +166,7 @@
 	(PCIE_CONF_BUS(bus) | PCIE_CONF_DEV(PCI_SLOT(devfn))	| \
 	 PCIE_CONF_FUNC(PCI_FUNC(devfn)) | PCIE_CONF_REG(where))
 
-#define PIO_TIMEOUT_MS			1
+#define PIO_TIMEOUT_MS			30
 
 #define LINK_WAIT_MAX_RETRIES		10
 #define LINK_WAIT_USLEEP_MIN		90000
-- 
2.17.1

